src/PlusDataCollection/OpenHaptics/vtkPlusOpenHapticsDevice.cxx

```diff
@@ -0,0 +1,297 @@
+#include <vtkMatrix4x4.h>
+#include <vtkTransform.h>
+
+// STL includes
+#include <fstream>
+#include <iostream>
+#include <sstream>
+
+vtkStandardNewMacro(vtkPlusOpenHapticsDevice);
+
+vtkPlusOpenHapticsDevice::vtkPlusOpenHapticsDevice()
+{
+  this->FrameNumber = 0;
+  this->isHapticDeviceInitialized = false;
+  this->DeviceHandle = -1;
+  this->RequirePortNameInDeviceSetConfiguration = true;
+  this->StartThreadForInternalUpdates = true;
+  this->AcquisitionRate = 20;
+  this->DeviceName = "PLUS";
+}
+
+vtkPlusOpenHapticsDevice::~vtkPlusOpenHapticsDevice()
+{
+  if (this->isHapticDeviceInitialized)
+  {
+    this->isHapticDeviceInitialized = false;
+  }
+}
+
+std::string vtkPlusOpenHapticsDevice::GetSdkVersion()
+{
+  std::stringstream ss;
+  ss << HD_VERSION_MAJOR_NUMBER << "." << HD_VERSION_MINOR_NUMBER << "." << HD_VERSION_BUILD_NUMBER;
+  return ss.str();
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::Probe()
+{
+  if (this->isHapticDeviceInitialized)
+  {
+    LOG_ERROR("vtkPlusOpenHapticsDevice::Probe should not be called while the device is already initialized");
+    return PLUS_FAIL;
+  }
+  if (this->InternalConnect())
+  {
+    this->InternalDisconnect();
+    LOG_INFO("vtkPlusOpenHapticsDevice::Probe found OpenHaptics Device");
+    return PLUS_SUCCESS;
+  }
+  else
+  {
+    LOG_ERROR("vtkPlusOpenHapticsDevice::Probe failed to connect to device");
+    return PLUS_FAIL;
+  }
+
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::InternalUpdate()
+{
+  //schedule the internal callback - must be done each time device is polled
+  //the scheduler is intended to manage multiple clients requesting data from the same device
+  hdScheduleSynchronous(positionCallback, this, HD_MAX_SCHEDULER_PRIORITY);
+
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::InternalConnect()
+{
+  if (this->isHapticDeviceInitialized)
+  {
+    LOG_DEBUG("Already connected - " << this->DeviceName);
+    return PLUS_SUCCESS;
+  }
+
+  if (this->InputChannels.empty())
+  {
+    LOG_WARNING("No force input has been provided");
+  }
+  else
+  {
+    if (this->InputChannels.size() > 1)
+    {
+      LOG_WARNING("Multiple input channels present.  Only first channel will be checked for force data");
+    }
+    vtkPlusDataSource* forceInput;
+    if (!this->InputChannels[0]->GetOwnerDevice()->GetToolByPortName("Force", forceInput))
+    {
+      LOG_WARNING("No Force tool in input channel.  Forces will not be available");
+    }
+  }
+
+  HDErrorInfo errorFlush;
+  while (HD_DEVICE_ERROR(errorFlush = hdGetError())) {}
+
+  DeviceHandle = hdInitDevice(this->DeviceName.c_str());
+
+  HDErrorInfo error;
+  if (HD_DEVICE_ERROR(error = hdGetError()))
+  {
+    LOG_ERROR( "Failed to initialize Phantom Omni " << this->DeviceName);
+    DeviceHandle = -1;
+    return PLUS_FAIL;
+  }
+
+  // Enable forces
+  hdEnable(HD_FORCE_OUTPUT);
+  hdEnable(HD_FORCE_RAMPING);
+
+  this->isHapticDeviceInitialized = true;
+  LOG_INFO("Phantom initialized: " << this->DeviceName)

```

Comment:
This may be more appropriate as LOG_DEBUG, however totally up to you, LOG_INFO is not wrong!